### PR TITLE
feat: Add validation support for pipeline parallelism

### LINKED ISSUE #578: Support validation step with PP
**Is your feature request related to a problem? Please describe.**
Currently validation step is skipped when PP > 1

**Describe the solution you'd like**
Support for validation

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by hemildesai**
closes #578

### CODE REVIEW

File: examples/llm_finetune/llama3_1/llama3_1_8b_hellaswag_pp.yaml [hunk 1]
```diff
@@ -37,6 +37,7 @@ rng:
 model:
   _target_: nemo_automodel.NeMoAutoModelForCausalLM.from_pretrained
   pretrained_model_name_or_path: meta-llama/Llama-3.1-8B
+  force_hf: true
```
Comment (akoumpa) [inline examples/llm_finetune/llama3_1/llama3_1_8b_hellaswag_pp.yaml:40]:
```suggestion
```

do we need this one?

File: nemo_automodel/recipes/llm/train_ft.py [hunk 2]
```diff
@@ -1260,10 +1244,18 @@ def _forward_backward_step(
...
+                        self.pp.info.schedule.step(input_ids, target=targets, losses=losses, **batch)
+                    else:
+                        self.pp.info.schedule.step(target=targets, losses=losses, **batch)
                 else:
-                    self.pp.info.schedule.step(target=targets, losses=losses, **batch)
+                    # Use eval for validation (forward only, no backward)
+                    if self.pp.info.has_first_stage:
+                        self.pp.info.schedule.eval(input_ids, target=targets, losses=losses, **batch)
+                    else:
+                        self.pp.info.schedule.eval(target=targets, losses=losses, **batch)
```
Comment (akoumpa) [inline nemo_automodel/recipes/llm/train_ft.py:1258]: this block (from if self.pp_enabled to here) is way too complicated IMHO.

Do you think it's possible to use `.step` alone and avoid having `.eval` to begin with?

PyTorch has a state-changing `.eval` method (in `nn.Module`) so I want to avoid the name/semantics conflict.

File: nemo_automodel/recipes/llm/train_ft.py [hunk 3]
```diff
@@ -1447,10 +1439,22 @@ def _run_validation_epoch(self, val_dataloader):
                 total_loss += torch.sum(torch.stack(loss_buffer)).item()
                 total_num_label_tokens += num_label_tokens
 
-        total_loss = self._dp_allreduce(total_loss, include_cp=True).item()
+        total_loss = self._dp_allreduce(total_loss, include_cp=True)
         total_num_label_tokens = self._dp_allreduce(torch.tensor(total_num_label_tokens, dtype=torch.long)).item()
```
Comment (akoumpa) [inline nemo_automodel/recipes/llm/train_ft.py:1443]: Thanks for this, would it make sense to remove `.item()` from `total_num_label_tokens` too?

File: nemo_automodel/recipes/llm/train_ft.py (hunk 2 above)
Comment (hemildesai) [reply nemo_automodel/recipes/llm/train_ft.py:1258]: Unfortunately as of now, the PP PyTorch API requires the use of `.eval` when gradients are not required.

Comment (akoumpa) [reply nemo_automodel/recipes/llm/train_ft.py:1258]: Thanks @hemildesai , I feel the PyTorch API choice is creating unnecessary complexity on the user side here.
